### PR TITLE
Python serializing solvers/models/prox

### PR DESCRIPTION
**Pull request by Dekken**
PR for tests

### CODE REVIEW

File: lib/include/tick/prox/prox_l1.h [hunk 1]
```diff
@@ -32,9 +26,13 @@ class DLL_PUBLIC TProxL1 : public TProxSeparable<T> {
   T value_single(T x) const override;
 
  public:
-  TProxL1(T strength, bool positive);
+  // This exists soley for cereal which has friend access
```
Comment (Mbompr) [inline lib/include/tick/prox/prox_l1.h]: ` which has friend access` is not meaningful anymore

File: lib/include/tick/prox/prox_separable.h [hunk 2]
```diff
@@ -82,9 +83,9 @@ class DLL_PUBLIC TProxSeparable : public TProx<T> {
 using ProxSeparable = TProxSeparable<double>;
 
 using ProxSeparableDouble = TProxSeparable<double>;
-CEREAL_REGISTER_TYPE(ProxSeparableDouble)
+// CEREAL_REGISTER_TYPE(ProxSeparableDouble)
```
Comment (Mbompr) [inline lib/include/tick/prox/prox_separable.h]: Left comments

File: lib/include/tick/solver/sto_solver.h [hunk 3]
```diff
@@ -9,7 +9,11 @@
 
 #include "tick/base/base.h"
 #include "tick/base_model/model.h"
+
+#include "tick/linear_model/model_logreg.h"
```
Comment (Mbompr) [inline lib/include/tick/solver/sto_solver.h]: Why do we need to include logreg and proxl1?

File: lib/swig/prox/prox_tv.i [hunk 4]
```diff
@@ -14,14 +14,15 @@ class TProxTV : public TProx<T> {
...
              unsigned long end,
              bool positive);
-};
 
-%template(ProxTV) TProxTV<double>;
-typedef TProxTV<double> ProxTV;
+  bool compare(const TProxTV<T> &that);
+};
 
 %template(ProxTVDouble) TProxTV<double>;
```
Comment (Mbompr) [inline lib/swig/prox/prox_tv.i]: Can you remind me when we can use `%template` and when we must use `%rename` ?

File: tick/linear_model/tests/serializing_test.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
+
+# License: BSD 3 clause
+
+
+import io, unittest
+import numpy as np
+
+import dill as pickle
```
Comment (Mbompr) [inline tick/linear_model/tests/serializing_test.py]: Cant' we use pickle directly?

File: tick/linear_model/tests/serializing_test.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+        }
+
+        for mod in model_map:
+            model = mod(fit_intercept=False)
+
+            coeffs0 = weights_sparse_gauss(20, nnz=5)
+            interc0 = None
+
+            features, labels = model_map[mod](
+                coeffs0, interc0, n_samples=3000, verbose=False,
```
Comment (Mbompr) [inline tick/linear_model/tests/serializing_test.py]: Is this faster with `n_samples=100`? Because it would already be way more than enough

File: tick/linear_model/tests/serializing_test.py [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+from tick.prox import ProxL1
+from tick.linear_model import ModelLinReg, SimuLinReg
+from tick.linear_model import ModelLogReg, SimuLogReg
+from tick.linear_model import ModelPoisReg, SimuPoisReg
+from tick.linear_model import ModelHinge, ModelQuadraticHinge, ModelSmoothedHinge
+
+from tick.simulation import weights_sparse_gauss
+
+class Test(TestSolver):
+    def test_solver_saga(self):
```
Comment (Mbompr) [inline tick/linear_model/tests/serializing_test.py]: change test name and add quick description in docstring

File: tick/robust/tests/serializing_test.py [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+from tick.linear_model import ModelPoisReg, SimuPoisReg
+from tick.linear_model import ModelHinge, ModelQuadraticHinge, ModelSmoothedHinge
+
+from tick.robust import ModelAbsoluteRegression, ModelEpsilonInsensitive, ModelHuber, \
+                        ModelLinRegWithIntercepts, ModelModifiedHuber
+
+from tick.simulation import weights_sparse_gauss
+
+class Test(TestGLM):
+    def test_solver_saga(self):
```
Comment (Mbompr) [inline tick/robust/tests/serializing_test.py]: change test name and add quick description in docstring

File: tick/robust/tests/serializing_test.py [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+        }
+
+        for mod in model_map:
+            model = mod(fit_intercept=False)
+
+            coeffs0 = weights_sparse_gauss(20, nnz=5)
+            interc0 = None
+
+            features, labels = model_map[mod](
+                coeffs0, interc0, n_samples=3000, verbose=False,
```
Comment (Mbompr) [inline tick/robust/tests/serializing_test.py]: same here for n_samples

File: tick/robust/tests/serializing_test.py [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+                seed=123).simulate()
+            model.fit(features, labels)
+
+            pickled = pickle.loads(pickle.dumps(model))
+
+            self.assertTrue(model._model.compare(pickled._model))
```
Comment (Mbompr) [inline tick/robust/tests/serializing_test.py]: Maybe can we also check that `model.loss(features[0]) == pickled.loss(features[0])`

File: tick/solver/tests/serializing_test.py [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+
+from tick.robust import ModelAbsoluteRegression, ModelEpsilonInsensitive, ModelHuber, \
+                        ModelLinRegWithIntercepts, ModelModifiedHuber
+
+from tick.solver import AdaGrad, SGD, SDCA, SAGA, SVRG
+
+from tick.simulation import weights_sparse_gauss
+
+class Test(TestSolver):
+    def test_solver_saga(self):
```
Comment (Mbompr) [inline tick/solver/tests/serializing_test.py]: same here

File: tick/solver/tests/serializing_test.py [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+            ModelLogReg : SimuLogReg,
+            ModelPoisReg: SimuPoisReg,
+            ModelHinge  : SimuLogReg,
+            ModelQuadraticHinge : SimuLogReg,
+            ModelSmoothedHinge : SimuLogReg,
+
+            ModelAbsoluteRegression : SimuLinReg,
+            ModelEpsilonInsensitive : SimuLinReg,
+            ModelHuber              : SimuLinReg,
+            # ModelLinRegWithIntercepts : SimuLinReg, # has coeffe issue?
```
Comment (Mbompr) [inline tick/solver/tests/serializing_test.py]: See https://x-datainitiative.github.io/tick/modules/robust.html#example to have an example of data simulated for `ModelLinRegWithIntercepts`

File: lib/include/tick/prox/prox_l1.h (hunk 1 above)
Comment (Dekken) [reply lib/include/tick/prox/prox_l1.h]: this is true!

File: lib/include/tick/solver/sto_solver.h (hunk 3 above)
Comment (Dekken) [reply lib/include/tick/solver/sto_solver.h]: leftovers likely - to remove

File: lib/swig/prox/prox_tv.i (hunk 4 above)
Comment (Dekken) [reply lib/swig/prox/prox_tv.i]: In so far as I have found

template is fine if there is no complex types in parameter declarations - anywhere I have had to use "array<double> or array<float>" has failed with "array<T>" and so expansion using "rename" was found to work

File: tick/linear_model/tests/serializing_test.py (hunk 5 above)
Comment (Dekken) [reply tick/linear_model/tests/serializing_test.py]: yep

File: tick/robust/tests/serializing_test.py (hunk 10 above)
Comment (Dekken) [reply tick/robust/tests/serializing_test.py]: k

File: tick/solver/tests/serializing_test.py (hunk 12 above)
Comment (Dekken) [reply tick/solver/tests/serializing_test.py]: k

File: tick/linear_model/tests/serializing_test.py (hunk 5 above)
Comment (Mbompr) [reply tick/linear_model/tests/serializing_test.py]: Yep means we can or we can't?

Comment (Dekken) [reply tick/linear_model/tests/serializing_test.py]: Yep means I did but forgot one

Comment (Dekken) [reply tick/linear_model/tests/serializing_test.py]: maybe you need to refresh?

Comment (Mbompr) [reply tick/linear_model/tests/serializing_test.py]: Oh ok sorry, it's indeed shown as outdated

Comment (Dekken) [reply tick/linear_model/tests/serializing_test.py]: can and updated
